Zap waiting for Spider endlessly when it crashed.

I have an example reposity that has ZAP waiting for the Spider to end at 96% done. The spider never ends because apparently it crashes. Hitting CTRL-C displays an exception. In this case it is a HSQLDB problem:
`org.hsqldb.HsqlException: Data cache size limit is reached: 10000`

The issue is that ZAP doesn't detect this and just waits for the spider forever instead of failling or displaying an error.

Reproduction steps:

```
git clone https://gitlab.com/Flockademic/Flockademic.git
cd Flockademic
git checkout 1e6dbaf6 
mkdir wrk
docker run -ti --rm -v $PWD:/zap/wrk owasp/zap2docker-live zap-baseline.py -d -t https://flockademic.com/ 
```
It will start the spider and it quickly will get stuck at 96% . Interrupt with CTRL-c to see the exception. Here it is :

```
11512 [ZAP-SpiderInitThread-0] INFO org.zaproxy.zap.extension.spider.SpiderThread  - Starting spidering scan on https://flockademic.com at Thu Feb 22 13:29:31 UTC 2018
11514 [ZAP-SpiderInitThread-0] INFO org.zaproxy.zap.spider.Spider  - Spider initializing...
11550 [ZAP-SpiderInitThread-0] INFO org.zaproxy.zap.spider.Spider  - Starting spider...
27227 [ZAP-SpiderThreadPool-0-thread-1] INFO hsqldb.db..ENGINE  - dataFileCache CACHE SIZE limit reached
27228 [ZAP-SpiderThreadPool-0-thread-1] FATAL hsqldb.db..ENGINE  - /home/zap/.ZAP_D/session/untitled1.data getFromFile failed 37817
org.hsqldb.HsqlException: Data cache size limit is reached: 10000
        at org.hsqldb.error.Error.error(Unknown Source)
        at org.hsqldb.error.Error.error(Unknown Source)
        at org.hsqldb.persist.Cache.put(Unknown Source)
        at org.hsqldb.persist.DataFileCache.getFromFile(Unknown Source)
        at org.hsqldb.persist.DataFileCache.get(Unknown Source)
        at org.hsqldb.persist.RowStoreAVLDisk.get(Unknown Source)
        at org.hsqldb.index.NodeAVLDisk.findNode(Unknown Source)
        at org.hsqldb.index.NodeAVLDisk.getRight(Unknown Source)
        at org.hsqldb.index.NodeAVLDisk.child(Unknown Source)
        at org.hsqldb.index.IndexAVL.insert(Unknown Source)
        at org.hsqldb.persist.RowStoreAVL.indexRow(Unknown Source)
        at org.hsqldb.persist.RowStoreAVLDisk.indexRow(Unknown Source)
        at org.hsqldb.TransactionManager2PL.addInsertAction(Unknown Source)
        at org.hsqldb.Session.addInsertAction(Unknown Source)
        at org.hsqldb.Table.insertSingleRow(Unknown Source)
        at org.hsqldb.StatementDML.insertSingleRow(Unknown Source)
        at org.hsqldb.StatementInsert.getResult(Unknown Source)
        at org.hsqldb.StatementDMQL.execute(Unknown Source)
        at org.hsqldb.Session.executeCompiledStatement(Unknown Source)
        at org.hsqldb.Session.execute(Unknown Source)
        at org.hsqldb.jdbc.JDBCPreparedStatement.fetchResult(Unknown Source)
        at org.hsqldb.jdbc.JDBCPreparedStatement.executeUpdate(Unknown Source)
        at org.parosproxy.paros.db.paros.ParosTableHistory.write(ParosTableHistory.java:350)
        at org.parosproxy.paros.db.paros.ParosTableHistory.write(ParosTableHistory.java:295)
        at org.parosproxy.paros.model.HistoryReference.<init>(HistoryReference.java:360)
        at org.zaproxy.zap.extension.spider.SpiderThread.notifySpiderTaskResult(SpiderThread.java:488)
        at org.zaproxy.zap.spider.Spider.notifyListenersSpiderTaskResult(Spider.java:809)
        at org.zaproxy.zap.spider.SpiderTask.run(SpiderTask.java:241)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
27229 [ZAP-SpiderThreadPool-0-thread-1] ERROR org.zaproxy.zap.extension.spider.SpiderThread  - java.sql.SQLException: Data cache size limit is reached: 10000
org.parosproxy.paros.db.DatabaseException: java.sql.SQLException: Data cache size limit is reached: 10000
        at org.parosproxy.paros.db.paros.ParosTableHistory.write(ParosTableHistory.java:297)
        at org.parosproxy.paros.model.HistoryReference.<init>(HistoryReference.java:360)
        at org.zaproxy.zap.extension.spider.SpiderThread.notifySpiderTaskResult(SpiderThread.java:488)
        at org.zaproxy.zap.spider.Spider.notifyListenersSpiderTaskResult(Spider.java:809)
        at org.zaproxy.zap.spider.SpiderTask.run(SpiderTask.java:241)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.sql.SQLException: Data cache size limit is reached: 10000
        at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
        at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
        at org.hsqldb.jdbc.JDBCPreparedStatement.fetchResult(Unknown Source)
        at org.hsqldb.jdbc.JDBCPreparedStatement.executeUpdate(Unknown Source)
        at org.parosproxy.paros.db.paros.ParosTableHistory.write(ParosTableHistory.java:350)
        at org.parosproxy.paros.db.paros.ParosTableHistory.write(ParosTableHistory.java:295)
        ... 7 more
Caused by: org.hsqldb.HsqlException: Data cache size limit is reached: 10000
        at org.hsqldb.error.Error.error(Unknown Source)
        at org.hsqldb.error.Error.error(Unknown Source)
        at org.hsqldb.persist.Cache.put(Unknown Source)
        at org.hsqldb.persist.DataFileCache.getFromFile(Unknown Source)
        at org.hsqldb.persist.DataFileCache.get(Unknown Source)
        at org.hsqldb.persist.RowStoreAVLDisk.get(Unknown Source)
        at org.hsqldb.index.NodeAVLDisk.findNode(Unknown Source)
        at org.hsqldb.index.NodeAVLDisk.getRight(Unknown Source)
        at org.hsqldb.index.NodeAVLDisk.child(Unknown Source)
        at org.hsqldb.index.IndexAVL.insert(Unknown Source)
        at org.hsqldb.persist.RowStoreAVL.indexRow(Unknown Source)
        at org.hsqldb.persist.RowStoreAVLDisk.indexRow(Unknown Source)
        at org.hsqldb.TransactionManager2PL.addInsertAction(Unknown Source)
        at org.hsqldb.Session.addInsertAction(Unknown Source)
        at org.hsqldb.Table.insertSingleRow(Unknown Source)
        at org.hsqldb.StatementDML.insertSingleRow(Unknown Source)
        at org.hsqldb.StatementInsert.getResult(Unknown Source)
        at org.hsqldb.StatementDMQL.execute(Unknown Source)
        at org.hsqldb.Session.executeCompiledStatement(Unknown Source)
        at org.hsqldb.Session.execute(Unknown Source)
        ... 11 more
27230 [ZAP-SpiderThreadPool-0-thread-1] ERROR org.zaproxy.zap.ZAP$UncaughtExceptionLogger  - Exception in thread "ZAP-SpiderThreadPool-0-thread-1"
java.lang.NullPointerException
        at org.zaproxy.zap.extension.spider.SpiderScan.notifySpiderTaskResult(SpiderScan.java:339)
        at org.zaproxy.zap.spider.Spider.notifyListenersSpiderTaskResult(Spider.java:809)
        at org.zaproxy.zap.spider.SpiderTask.run(SpiderTask.java:241)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

Increasing the cache size allow the run to complete, but the underlying issue is still there.

```
docker run -ti --rm -v $PWD:/zap/wrk/:rw owasp/zap2docker-live sh -c 'sed -i "s/SET FILES CACHE SIZE 10000/SET FILES CACHE SIZE 100000/" db/zapdb.script && zap-baseline.py -d -t https://flockademic.com/ '
```

--- a/src/org/zaproxy/zap/spider/SpiderTask.java
+++ b/src/org/zaproxy/zap/spider/SpiderTask.java
@@ -173,23 +173,29 @@ public class SpiderTask implements Runnable {
 
 	@Override
 	public void run() {
-		if (reference == null) {
-			log.warn("Null URI. Skipping crawling task: " + this);
+		try {
+			if (reference == null) {
+				log.warn("Null URI. Skipping crawling task: " + this);
+				return;
+			}
+
+			if (log.isDebugEnabled()) {
+				log.debug("Spider Task Started. Processing uri at depth " + depth
+						+ " using already constructed message: " + reference.getURI());
+			}
+
+			runImpl();
+		} finally {
 			parent.postTaskExecution();
-			return;
+			log.debug("Spider Task finished.");
 		}
-
-		// Log the task start
-		if (log.isDebugEnabled()) {
-			log.debug("Spider Task Started. Processing uri at depth " + depth
-					+ " using already constructed message:  " + reference.getURI());
-		}
-
+	}
+	
+	private void runImpl() {
 		// Check if the should stop
 		if (parent.isStopped()) {
 			log.debug("Spider process is stopped. Skipping crawling task...");
 			deleteHistoryReference();
-			parent.postTaskExecution();
 			return;
 		}
 
@@ -202,7 +208,6 @@ public class SpiderTask implements Runnable {
 			msg = prepareHttpMessage();
 		} catch (Exception e) {
 			log.error("Failed to prepare HTTP message: ", e);
-			parent.postTaskExecution();
 			return;
 		}
 
@@ -211,9 +216,6 @@ public class SpiderTask implements Runnable {
 		} catch (Exception e) {
 			setErrorResponse(msg, e);
 			parent.notifyListenersSpiderTaskResult(new SpiderTaskResult(msg, getSkippedMessage("ioerror")));
-
-			// The exception was already logged, in fetchResource, with the URL (which we dont have here)
-			parent.postTaskExecution();
 			return;
 		}
 
@@ -221,7 +223,6 @@ public class SpiderTask implements Runnable {
 		if (parent.isStopped()) {
 		    parent.notifyListenersSpiderTaskResult(new SpiderTaskResult(msg, getSkippedMessage("stopped")));
 			log.debug("Spider process is stopped. Skipping crawling task...");
-			parent.postTaskExecution();
 			return;
 		}
 		// Check if the crawling process is paused
@@ -239,7 +240,6 @@ public class SpiderTask implements Runnable {
 				}
 
 				parent.notifyListenersSpiderTaskResult(new SpiderTaskResult(msg, filterResult.getReason()));
-				parent.postTaskExecution();
 				return;
 			}
 		}
@@ -248,7 +248,6 @@ public class SpiderTask implements Runnable {
 		if (parent.isStopped()) {
 			parent.notifyListenersSpiderTaskResult(new SpiderTaskResult(msg, getSkippedMessage("stopped")));
 			log.debug("Spider process is stopped. Skipping crawling task...");
-			parent.postTaskExecution();
 			return;
 		}
 		// Check if the crawling process is paused
@@ -261,10 +260,6 @@ public class SpiderTask implements Runnable {
 		} else {
 			parent.notifyListenersSpiderTaskResult(new SpiderTaskResult(msg, getSkippedMessage("maxdepth")));
 		}
-
-		// Update the progress and check if the spidering process should stop
-		parent.postTaskExecution();
-		log.debug("Spider Task finished.");
 	}
 
 	private String getSkippedMessage(String key) {

